### ISSUE TITLE
Export to CoreML with dynamic shapes fails

### ISSUE THREAD

**Issue by msluszniak** [pytorch/executorch#4818]
### 🐛 Describe the bug

I tried to used dynamic shapes when exporting to Executorch with CoreML backend. Here is the example code to reproduce the error (based on export to CoreML script from this repo). I know that CoreML was not tested with Torch higher than 2.3.0, but I need to use the newer one, so if the problem comes from this fact it will be valuable information

```py
import torch
import executorch.exir as exir
from executorch.exir import to_edge
from torch.export import export, Dim
from executorch.exir.backend.backend_api import to_backend
from executorch.backends.apple.coreml.compiler import CoreMLBackend
import coremltools as ct
import copy
import argparse

_EDGE_COMPILE_CONFIG = exir.EdgeCompileConfig(
    _check_ir_validity=False,
    _skip_dim_order=True,
)

def parse_args() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-m",
        "--model_name",
        required=False
    )

    parser.add_argument(
        "-c",
        "--compute_unit",
        required=False,
        default=ct.ComputeUnit.ALL.name.lower(),
    )

    parser.add_argument(
        "-precision",
        "--compute_precision",
        required=False,
        default=ct.precision.FLOAT16.value,
    )

    parser.add_argument(
        "--compile",
        action=argparse.BooleanOptionalAction,
        required=False,
        default=False,
    )
    parser.add_argument("--use_partitioner", action=argparse.BooleanOptionalAction)
    parser.add_argument("--generate_etrecord", action=argparse.BooleanOptionalAction)
    parser.add_argument("--save_processed_bytes", action=argparse.BooleanOptionalAction)

    args = parser.parse_args()
    return args

def generate_compile_specs_from_args(args):
    model_type = CoreMLBackend.MODEL_TYPE.MODEL
    if args.compile:
        model_type = CoreMLBackend.MODEL_TYPE.COMPILED_MODEL

    compute_precision = ct.precision(args.compute_precision)
    compute_unit = ct.ComputeUnit[args.compute_unit.upper()]

    return CoreMLBackend.generate_compile_specs(
        compute_precision=compute_precision,
        compute_unit=compute_unit,
        model_type=model_type,
    )

def lower_module_to_coreml(module, compile_specs, dynamic_shapes=None):
    with torch.no_grad():
        module = module.eval()
        if dynamic_shapes is None:
            edge = to_edge(export(module, example_inputs, ), compile_config=_EDGE_COMPILE_CONFIG)
        else:
            edge = to_edge(export(module, example_inputs, dynamic_shapes=dynamic_shapes), compile_config=_EDGE_COMPILE_CONFIG)
        edge_copy = copy.deepcopy(edge)
        lowered_module = to_backend(
            CoreMLBackend.__name__,
            edge.exported_program(),
            compile_specs,
        )

        return lowered_module, edge_copy


def export_lowered_module_to_executorch_program(lowered_module, example_inputs, dynamic_shapes=None):
    with torch.no_grad():
        lowered_module(*example_inputs)
        if dynamic_shapes is None:
            exec_prog = to_edge(
                export(lowered_module, example_inputs), compile_config=_EDGE_COMPILE_CONFIG
            ).to_executorch(
                config=exir.ExecutorchBackendConfig(
                    extract_constant_segment=False, extract_delegate_segments=True
                )
            )
        else:
            exec_prog = to_edge(
                export(lowered_module, example_inputs, dynamic_shapes=dynamic_shapes), compile_config=_EDGE_COMPILE_CONFIG
            ).to_executorch(
                config=exir.ExecutorchBackendConfig(
                    extract_constant_segment=False, extract_delegate_segments=True
                )
            )
        return exec_prog

class TestModule:
    def __init__(self):
        self.model_type = "test_module"
            
    def get_model(self) -> torch.nn.Module:  
        class Basic(torch.nn.Module):
            def __init__(self):
                super().__init__()

            def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
                return x + y
        return Basic()
            
    def get_example_inputs(self):
        return (torch.randn(3, 3), torch.randn(3, 3))


if __name__ == "__main__":
    args = parse_args()
    prompt_encoder = TestModule()
    model = prompt_encoder.get_model()
    example_inputs = prompt_encoder.get_example_inputs()

    compile_specs = generate_compile_specs_from_args(args)
    lowered_module = None

    dim1_x = Dim("dim1_x", min=1, max=10)
    dynamic_shapes = {"x": {1: dim1_x}, "y": {1: dim1_x}}

    # it fails here
    lowered_module, edge_copy = lower_module_to_coreml(
        module=model,
        compile_specs=compile_specs,
        dynamic_shapes=dynamic_shapes
    )
    
    exec_program = export_lowered_module_to_executorch_program(
        lowered_module,
        example_inputs,
        dynamic_shapes=dynamic_shapes
    )
```

It raises the following error:

```
Torch version 2.5.0.dev20240716 has not been tested with coremltools. You may run into unexpected errors. Torch 2.3.0 is the most recent version that has been tested.
Traceback (most recent call last):
  File "~/executorch/test_executorch.py", line 134, in <module>
    lowered_module, edge_copy = lower_module_to_coreml(
                                ^^^^^^^^^^^^^^^^^^^^^^^
  File "~/executorch/test_executorch.py", line 74, in lower_module_to_coreml
    lowered_module = to_backend(
                     ^^^^^^^^^^^
  File "~/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/executorch/exir/backend/backend_api.py", line 114, in _
    preprocess_result: PreprocessResult = cls.preprocess(
                                          ^^^^^^^^^^^^^^^
  File "~/site-packages/executorch/backends/apple/coreml/compiler/coreml_preprocess.py", line 384, in preprocess
    mlmodel = ct.convert(
              ^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/_converters_entry.py", line 635, in convert
    mlmodel = mil_convert(
              ^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/converter.py", line 188, in mil_convert
    return _mil_convert(model, convert_from, convert_to, ConverterRegistry, MLModel, compute_units, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/converter.py", line 212, in _mil_convert
    proto, mil_program = mil_convert_to_proto(
                         ^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/converter.py", line 288, in mil_convert_to_proto
    prog = frontend_converter(model, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/converter.py", line 108, in __call__
    return load(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/frontend/torch/load.py", line 74, in load
    converter = TorchConverter(
                ^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/frontend/torch/converter.py", line 586, in __init__
    self.graph = InternalTorchIRGraph.from_exir(exir=loaded_model)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/frontend/torch/internal_graph.py", line 524, in from_exir
    extract_inputs_from_exir_program(exported_program)
  File "~/site-packages/coremltools/converters/mil/frontend/torch/exir_utils.py", line 89, in extract_inputs_from_exir_program
    user_inputs.append(to_coreml_tensor_type(node.name, val))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/frontend/torch/exir_utils.py", line 18, in to_coreml_tensor_type
    return ct.TensorType(name=name, dtype=torch_to_mil_types[tensor.dtype], shape=tensor.shape)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/input_types.py", line 208, in __init__
    super(TensorType, self).__init__(name, shape)
  File "~/site-packages/coremltools/converters/mil/input_types.py", line 74, in __init__
    self.shape = _get_shaping_class(shape)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/site-packages/coremltools/converters/mil/input_types.py", line 590, in _get_shaping_class
    raise ValueError("Can't convert to CoreML shaping class from {}.".format(shape))
ValueError: Can't convert to CoreML shaping class from torch.Size([3, s0]).
```

### Versions

Collecting environment information...
PyTorch version: 2.5.0.dev20240716
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.30.1
Libc version: N/A

Python version: 3.11.0 (main, Mar  1 2023, 12:33:14) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] executorch==0.4.0a0+1114539
[pip3] executorchcoreml==0.0.1
[pip3] numpy==1.23.2
[pip3] torch==2.5.0.dev20240716
[pip3] torchaudio==2.4.0.dev20240716
[pip3] torchsr==1.0.4
[pip3] torchvision==0.20.0.dev20240716
[conda] executorch                0.4.0a0+1114539          pypi_0    pypi
[conda] executorchcoreml          0.0.1                    pypi_0    pypi
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] torch                     2.5.0.dev20240716          pypi_0    pypi
[conda] torchaudio                2.4.0.dev20240716          pypi_0    pypi
[conda] torchsr                   1.0.4                    pypi_0    pypi
[conda] torchvision               0.20.0.dev20240716          pypi_0    pyp

**Comment by digantdesai** [comment]
Thanks for the detailed report.

@cccclai do we support dynamic shapes with coreml?

**Comment by cccclai** [comment]
Enable dynamic shape is an ongoing effort. Refer to https://github.com/pytorch/executorch/issues/4658

**Comment by msluszniak** [comment]
Thank you for quick response :)

**Comment by YifanShenSZ** [comment]
Hi @msluszniak, we have recently released coremltools 8.0b2, which adds a primitive support for dynamic shape. Please try it out by
```
pip install coremltools==8.0b2
```
Let us know if any new error comes out 🙏

**Comment by YifanShenSZ** [comment]
Glad coremltools 8.0b2 has resolved this issue 😆 Yes, please close this one and file a new [coremltools issue](https://github.com/apple/coremltools/issues).

Please create a minimal reproduce directly using coremltools convert workflow, though, since the to_backend workflow may complicate our investigation
```
 import torch 
 from torch.export import export
 from executorch.exir import to_edge
 
 import coremltools as ct
 
 example_args = (torch.randn(*size), )
 aten_dialect = export(AnyNNModule(), example_args)
 edge_dialect = to_edge(aten_dialect).exported_program()
 
 mlmodel = ct.convert(edge_dialect)
```
